noria-server/src/controller/migrate/mod.rs

```diff
@@ -451,24 +542,93 @@ impl<'a> Migration<'a> {
                 dns
             });
 
-        // Boot up new domains (they'll ignore all updates for now)
-        debug!(log, "booting new domains");
-        for domain in changed_domains {
-            if mainline.domains.contains_key(&domain) {
-                // this is not a new domain
-                continue;
+        // Bucket reader replica nodes and the non-replica nodes separately.
+        let mut replica_map = HashMap::new();
+        let mut other_nodes = HashSet::new();
+        for &&ni in sorted_new
+                .iter()
+                .filter(|&&&ni| ni != mainline.source)
+                .filter(|&&&ni| !mainline.ingredients[ni].is_dropped())
+                .into_iter() {
+            // Check if the node is a reader, and then insert it in the replica map. The value
+            // is a list of readers that are for the same node, and the key is the node the
+            // readers are for. Later, we will also add the associated ingress node to the replica
+            // map, since readers are all located on their own domains.
+            if let Some(for_node) = mainline.ingredients[ni]
+                    .with_reader(|r| Some(r.is_for()))
+                    .unwrap_or(None) {
+                if mainline.ingredients[ni].replica_index().is_some() {
+                    replica_map.entry(for_node).or_insert(HashSet::new());
+                    replica_map.get_mut(&for_node).unwrap().insert(ni);
+                }
+            } else {
+                other_nodes.insert(ni);
             }
+        }
 
-            let nodes = uninformed_domain_nodes.remove(&domain).unwrap();
-            let d = mainline.place_domain(
-                domain,
-                mainline.ingredients[nodes[0].0].sharded_by().shards(),
-                &log,
-                nodes,
+        // Obtain a vec of DomainIndexes in the order that we want to assign them to workers.
+        // For replicas, all replicas for the same reader should be consecutive. For non-replicas,
+        // it doesn't really matter unless we want the ordering to be deterministic.
+        let mut changed_domains_replicas = Vec::new();
+        for (_, readers) in &mut replica_map {
+            let mut domains = readers
+                .iter()
+                .filter(|&ni| !mainline.ingredients[*ni].is_dropped())
+                .map(|&ni| mainline.ingredients[ni].domain())
+                .collect::<Vec<_>>();
```

Comment:
Why do you need to `collect` here if you're then immediately `append`ing to `changed_domains_replicas`? Couldn't you just give the iterator directly to `.extend` instead?